### PR TITLE
ROX-18482: Implement notification converter

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR implements the notification converter.

The event converter's main responsibility is to convert a structured log entry to a `storage.Notification`.

For our current logging library we use zap. The main thing for the converter is that it has to convert the "context", i.e. the [zap.Field](https://pkg.go.dev/go.uber.org/zap/zapcore#Field)s to a `map[string]string`, which will later on be used to present to the user within the API / UI.

To correctly encode the fields to their appropriate type, a [ObjectEncoder](https://pkg.go.dev/go.uber.org/zap/zapcore#ObjectEncoder) had to be implemented. The object encoder's main responsiblity is to encode a field value to a string value (opposed to the existing encoders within zap, which either encode 
these to JSON values or `interface{}` values).

As an alternative, one could have also implemented the [Core](https://pkg.go.dev/go.uber.org/zap/zapcore#Core) interface, as this also has direct access to the field. However, that would have increased the scope of the implementation hence decided not to pursue this (at the very least, for the time being).

Few caveats:
The converter currently does the following:
- Map the zapcore.Fields to labes, i.e. a map[string]string.
- For each field, only send "resource relevant" fields as notification.
- Introduce hints / domains based on resources and module paths.

Not all fields that _can_ be set via `zap.Field` are currently converted. The reason is that the benefit of exposing those isn't necessarily clear to me at this current time(e.g. why would we want to expose an array of objects / a complete struct).
If we want to support those, we simply have to implement the interface accordingly.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- see unit tests.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #7257** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7257" target="_blank">Graphite</a> 
    * **PR #7423** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7423" target="_blank">Graphite</a> 
      * **PR #7424** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7424" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/7424?utm_source=stack-comment).

File: pkg/logging/encoder.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+func (z *stringObjectEncoder) AddInt64(key string, value int64) {
+	z.m[key] = strconv.FormatInt(value, 10)
+}
+
+func (z *stringObjectEncoder) AddString(key, value string) {
+	z.m[key] = value
+}
+
+func (z *stringObjectEncoder) AddTime(key string, value time.Time) {
+	z.m[key] = value.String()
```
Comment (stehessel) [inline pkg/logging/encoder.go:43]:
```suggestion
	z.m[key] = value.Format(time.RFC3339Nano)
```
Example:
```
.String() = 2000-02-01 12:13:14.000000015 +0000 UTC
.Format(...) = 2000-02-01T12:13:14.000000015Z
```

Although maybe personal taste I guess.

File: pkg/logging/fields.go [hunk 2]
```diff
@@ -9,20 +28,27 @@ func Err(err error) zap.Field {
...
 
 // NodeID provides the node ID as a structured log field.
 func NodeID(id string) zap.Field {
-	return zap.String("node_id", id)
+	return zap.String(nodeIDField, id)
+}
+
+// isResourceTypeField returns whether the given zap.Field is related to a resource.
+// If it is, it will return true and the name of the resource.
+func isResourceTypeField(field zap.Field) (bool, string) {
```
Comment (stehessel) [inline pkg/logging/fields.go:51]:
```suggestion
func getResourceTypeField(field zap.Field) (string, bool) {
```
Seems more golang idomatic to me.

File: pkg/logging/log_converter_impl.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	// For now, the assumption is that structured logging with our current logger uses the construct
+	// according to https://github.com/uber-go/zap/blob/master/field.go. Thus, the given interfaces
+	// shall be a strongly-typed zap.Field.
+	var resourceType string
+	var resourceTypeKey string
+	for _, c := range context {
+		// Currently silently drop the given context of the log entry if it's not a zap.Field.
+		if field, ok := c.(zap.Field); ok {
+			field.AddTo(enc)
+			if isResourceTypeField, resource := isResourceTypeField(field); isResourceTypeField {
```
Comment (stehessel) [inline pkg/logging/log_converter_impl.go:30]: Should we warn here if somebody passed two different resource fields?

File: pkg/logging/encoder.go (hunk 1 above)
Comment (dhaus67) [reply pkg/logging/encoder.go]: Just had a similar comment in another PR, seems like we settle on `Format(time.RFC3339)`.

**Review by stehessel — APPROVED** [review]

**Pull request merged**